### ISSUE TITLE
Refactor populate_course_network cis command to batch users

### ISSUE THREAD

**Issue by btopro** [elmsln/elmsln#1945]
Issue:
Very high enrollment courses / sections (1,500+) are currently causing issues finishing the sync in larger builds. This makes sense cause it's attempting to load, verify and resave / add that many people in a single pass, then move onto the next system.

resolution:
We assemble the roster per course and already do batching in this way to distribute load but we need to also put a cap on the number of users we'll process at a time. This would allow for batching within the context of a listing of course users. To do this, we either ship all the users and have each site only process 200 at a time before re-posting to itself, OR we do the batching ahead of time and let CIS broker this out.

alternate resolution:
Let's see if we can leverage a crazier part of httprl to multi-thread function processing. If we dropped into a multi-threaded approach to solving this by crunching the same function with MORE workers (sounds bad I know) that each then have their own process that has the ability to not timeout, we could actually have the intial thread keep moving forward while a non-block loop keeps churning through the current info. This was designed to keep long jobs from timing out.

Examples from httprl:

```

// List of nodes to load; 241-244.
$nodes = array(241 => '', 242 => '', 243 => '', 244 => '');
foreach ($nodes as $nid => &$node) {
  // Setup callback options array.
  $callback_options = array(
    array(
      'function' => 'node_load',
      'return' => &$node,
      // Setup options array.
      'options' => array(
        'domain_connections' => 2, // Only use 2 threads for this request.
      ),
    ),
    $nid,
  );
  // Queue up the request.
  httprl_queue_background_callback($callback_options);
}
// Execute request.
httprl_send_request();

// Echo what was returned.
echo httprl_pr($nodes);
?>



Load nodes 50-100 using httprl_batch_callback and node_load_multiple.

<?php
// List of nodes to load; 50-100.
$nids = range(50, 100);
// Run not parallel if background callbacks are disabled.
if (!httprl_is_background_callback_capable()) {
  $results = node_load_multiple($nids);
}
else {
  // Queue & Execute requests.
  $results = httprl_batch_callback('node_load_multiple', $nids);
}
// Echo what was returned.
echo httprl_pr($results);
?>



Load nodes 50-100 using httprl_batch_callback and node_load.

<?php
// List of nodes to load; 50-100.
$nids = range(50, 100);
// Run not parallel if background callbacks are disabled.
if (!httprl_is_background_callback_capable()) {
  // httprl_run_multiple does a foreach on $nids running every $value through
  // the given callback.
  $results = httprl_run_multiple('node_load', $nids);
}
else {
  // Set options.
  $options = array(
    'multiple_helper' => TRUE,
  );
  // Queue & Execute requests.
  $results = httprl_batch_callback('node_load', $nids, $options);
}
// Echo what was returned.
echo httprl_pr($results);
?>


Load nodes 50-100 using httprl_batch_callback and node_load_multiple as user 1.

<?php
// List of nodes to load; 50-100.
$nids = range(50, 100);
// Run not parallel if background callbacks are disabled.
if (!httprl_is_background_callback_capable()) {
  // Run node_load_multiple as user 1
  $current_account = $GLOBALS['user']->uid;
  $GLOBALS['user'] = user_load(1);
  $results = node_load_multiple($nids);
  // Set global user back.
  $GLOBALS['user'] = $current_account;
}
else {
  // Set options.
  $options = array(
    'context' => array(
      'uid' => 1,
    ),
  );
  // Queue & Execute requests.
  $results = httprl_batch_callback('node_load_multiple', $nids, $options);
}
// Echo what was returned.
echo httprl_pr($results);
?>



Run a function in the background. Notice that there is no return.

<?php
// Call watchdog('httprl-test', 'background watchdog call done');
// in the background/async.
httprl_call_user_func_array_async('watchdog', array('httprl-test', 'background watchdog call done 2'));

?>
```

**Comment by btopro** [comment]
decision point: does CIS broker the data out to everyone or is everyone responsible for brokering themselves. I think it's more optimal to do the data split once and have CIS basically call itself mutlitple times on the same data piece until it's finished digesting it (waiting 10 seconds or whatever after sending then calling itself w/ the next batch of info). This is less data sent across the network and less processing overall. so

- load courses A and B
- A has 50 students, ok just ship it
- B has 5000 students, break them into 250 at a time and send the first 250 at the next system
  - then take the remaining 4,750 and send them back at itself, re-invoking the call but passing this list of users back to digest the chunk
  - doesn't pop an item off the queue, it looks for a roster / flag for batchjob, skips loading up users / course info and lobs off the next 250 to process, ships them off
  - recursively calls itself w/ the 4,500 users and keeps doing so till the array is empty
  - once array is empty, it calls itself w/o the roster group so that it pops off the next course in the process

**Comment by btopro** [comment]
last commit at least ensures apache / php variable handoff doesn't get in the way via giant POSTs. RIght now if using php-fpm this will all (eventually) work though obviously we don't want to rely on insanely high timeout values to get the job done. Still need to implement the todos

**Comment by btopro** [comment]
breaking job up into a queue process seems like a better alternative to httprl theading

**Comment by andrewbelcher** [comment]
Looking at the code (`_cis_section_create_accounts($roster)`) it looks like `$roster` isn't always the result of a POST, but can also come from other places, e.g. a node (`_cis_section_activate_section($node)`)? Do any of those other places need to roster to be processed before it can continue with what's currently going on, or is it all OK to process in the queue behind the scenes?

I'm also assuming backwards compatibility will be important?

**Comment by btopro** [comment]
Backwards compat important, possibly adding a flag to this function for if it's in the batch mode or not. Here's what calls this function in the scheme of things to trip it off as part of the system sync.

https://github.com/elmsln/elmsln/blob/master/core/dslmcode/shared/drupal-7.x/modules/elmsln_contrib/cis_connector/modules/services/cis_course_service/callbacks/v1/sync_data_model.inc

https://github.com/elmsln/elmsln/blob/master/core/dslmcode/shared/drupal-7.x/modules/elmsln_contrib/cis_connector/modules/authorities/cis_course_authority/callbacks/v1/sync_data_model.inc

**Comment by andrewbelcher** [comment]
Currently, `$actual` is a list across all sections, but the archiving happens per section. Is this intentional or a bug?

**Comment by andrewbelcher** [comment]
Also, are we expecting mail/password to be updated if the user already exists (it doesn't currently)?

**Comment by rickhumphries** [comment]
Hi @andrewbelcher , in terms of the mail/password to be updated if the user already exists - that would actually be functionality that we would want. Would there be any performance considerations were that to happen? 
Would need to see what @btopro 's thoughts on that were too if it doesn't currently do that as there might be a reason. 
Thanks

**Comment by andrewbelcher** [comment]
@rickhumphries Once you're processing things via cron rather than all in one go, performance issues aren't really much of a consideration. This is going in as a progressive enhancement which may not be implemented by default, so is could have implications for existing sites, but the solution would be fairly simple - to switch to background processing. I'll get that put into the MR, we can always change our mind or adjust how it does it.

**Comment by btopro** [comment]
Oversight on password updating part.

The sync everything routing also runs when sections are deactivated via a cron job in CIS broadcasting a shutdown command the sections after they go past their active date.

**Comment by andrewbelcher** [comment]
Cool - so password updating is already in the MR.

I'm still a little unclear as to which calls are suitable for deferring to cron. Basically, if there call requires the changes to be made before it can continue with the request, or will need the changes to happen near instantly (e.g. for the next request), then deferring is not suitable.

These are the calls I'm aware of, which of these *is* suitable for deferring:

- [ ] `cis_course_authority_elmsln_api_callback_sync_data_model()` (deferring in PR)
- [ ] `_cis_section_activate_section()` (not deferring in PR)
- [ ] `_cis_section_sync_roster()` (not deferring in PR)
- [ ] `cis_course_service_elmsln_api_callback_sync_data_model()` (deferring in PR)

**Comment by btopro** [comment]
`cis_course_authority_elmsln_api_callback_sync_data_model` and `cis_course_service_elmsln_api_callback_sync_data_model` are kicked off during a full cron or when an operation requires a sync from CIS (like a new section is added or updated then it'll execute this with only this course instead of all courses)

`_cis_section_sync_roster` I believe is legacy bc of the data model syncing parts as I can't trace it to firing anywhere (a year ago node sites synced themselves, now CIS brokers this transaction for all)

`_cis_section_activate_section` fires when a data model sync executes but ALSO in rare situations where individual sections are directly touched (like telling section A that it uses a different course structure then B which is then a direct save against just that node). In these cases it can also be sent into the background without being a big deal since we have other ways of triggering a sync

**Comment by andrewbelcher** [comment]
So should we do background processing for everything if it has been enabled site wide? The code will always process a batch first, so if you configure a batch size larger than the smaller operations, they would always be processed synchronously anyway.

**Comment by btopro** [comment]
yeah I'd say everything to ensure it goes through. When kicking off the job manually from CIS via http://online.elmsln.local/admin/config/system/cis/sync-roster (its on the status page) then it sends a job into the background thread anyway (as does cron to help avoid timeout issues). So if those BG threads were doing batch processing that'd be even better (and do so as the default setting).

**Comment by btopro** [comment]
- [ ] define a sane default batch size, internal tests stop at 3000 in a run. maybe 1000
- [ ] need to modify our snake routine in the event a batch doesn't fire
- [ ] try to eat tail natively on these batch routines; set the queue and then fire call back at self to resume on a new thread that can last longer
- [ ] test,test,test

**Comment by btopro** [comment]
Services working! Need to figure out the deal with authorities!!

**Comment by btopro** [comment]
This has been merged into the 0.8.1 release which was cut from master (so it's now master). I need to do additional testing against a 0.9.x variant before I merge there but looked good from testing!

**Comment by btopro** [comment]
iniital testing good; need to spin up a fresh VM running the merge and try and do the monster sync / from multiple sources and see how it goes
